db/migrate/20180131214527_create_submissions.rb

```diff
@@ -0,0 +1,12 @@
+class CreateSubmissions < ActiveRecord::Migration[5.1]
+  def change
+    create_table :submissions do |t|
+      t.string :title
+      t.string :url
+      t.text :text
+      t.belongs_to :user, index: true, null: false
+      t.references :post, index: true
+      t.timestamps
+    end
```

Comment:
I think this is a good start to simply treating "comments" as text posts on another submission. Let's say a feature request comes in that says "I want to show the total number of comments next to a post" (not unlike the "66 comments" in the following screenshot).

![screen shot 2018-02-06 at 9 01 34 am](https://user-images.githubusercontent.com/250031/35863356-6af91354-0b1c-11e8-97e1-a0e6f708837e.png)

How would you handle that with this data model?

Reply:
My first thought is to add a column for comment_id, similar to how I set up post_id. If comment_id is nil, then it is a comment directly in response to the post; if the comment_id has a number, it is a nested comment in reply to a parent comment. 

I'd also like to have an index that shows the sequence of the nested comments - as in first, second, third, or etc nested comment (indicated by an index value), under parent comment (indicated by comment_id). I've not used an index in this way before so I'm not sure if it can be executed the way I'm thinking.

If I can set up an index to indicate the sequence of nested comments relative to a parent comment, it would be good to also set up an index that tracks the sequence of non-nested comments relative to a parent post.

Reply:
> My first thought is to add a column for comment_id, similar to how I set up post_id. If comment_id is nil, then it is a comment directly in response to the post; if the comment_id has a number, it is a nested comment in reply to a parent comment.

That's a good thought, but I'm thinking of how would the view look like if you just want to iterate through immediate comments instead of all comments on a post (including nested)?  Probably with some sort of scope like the following?

```ruby
class Submission < ApplicationRecord
  has_many :direct_comments,
           ~> { where(comment_id: nil) },
           class_name: 'Submission',
           foreign_key: :post_id
end
```

Then in the view, you'd loop over them like:

```erb
<%= post.direct_comments.each do |comment| %>
  <%= render comment %>
<% end %>
```

That relationship and scope only works for the most-parent submission (a Post), but breaks down on comment submissions.

My thought on this would be to rename the `post_id` column to a more generic `parent_id` with a foreign key constraint setup on `submissions.id`. Then add a `counter_cache` column to keep track of the number of comments submitted to a post. You can read more about counter caches [here](http://guides.rubyonrails.org/association_basics.html) (search for `4.1.2.3 :counter_cache`). If you still want to proceed forward with the `post_id` and `comment_id` columns, then that says to me that these should be 2 separate tables (`posts` and `comments`) to better encapsulate the business logic.

> I'd also like to have an index that shows the sequence of the nested comments - as in first, second, third, or etc nested comment (indicated by an index value), under parent comment (indicated by comment_id). I've not used an index in this way before so I'm not sure if it can be executed the way I'm thinking.

> If I can set up an index to indicate the sequence of nested comments relative to a parent comment, it would be good to also set up an index that tracks the sequence of non-nested comments relative to a parent post.

That sounds cool. I've handled this in the past with a simple `position` column on the table. You can either pull in a gem, write a PostgreSQL sequence, or roll your own Ruby code to handle incrementing and updating that column as comments are added / removed.  Honestly though, the simplest approach would be to sort by the `created_at` timestamp with possibly the most recent first. Once voting gets into place, perhaps we'll consider sorting by "most upvoted".

Hope that helps!